Userland/Libraries/LibJS/Parser.cpp

```diff
@@ -2358,6 +2372,64 @@ NonnullRefPtr<ConditionalExpression> Parser::parse_conditional_expression(Nonnul
     return create_ast_node<ConditionalExpression>({ m_state.current_token.filename(), rule_start.position(), position() }, move(test), move(consequent), move(alternate));
 }
 
+NonnullRefPtr<OptionalChain> Parser::parse_optional_chain(NonnullRefPtr<Expression> base)
+{
+    auto rule_start = push_start();
+    Vector<OptionalChain::Reference> chain;
+    do {
+        if (match(TokenType::QuestionMarkPeriod)) {
+            consume(TokenType::QuestionMarkPeriod);
+            switch (m_state.current_token.type()) {
+            case TokenType::ParenOpen:
+                chain.append(OptionalChain::Call { parse_arguments(), OptionalChain::Mode::Optional });
+                break;
+            case TokenType::BracketOpen:
+                consume();
+                chain.append(OptionalChain::ComputedReference { parse_expression(2), OptionalChain::Mode::Optional });
+                consume(TokenType::BracketClose);
+                break;
+            case TokenType::TemplateLiteralStart:
+                // 13.3.1.1 - Static Semantics: Early Errors
+                // OptionalChain :
+                //        ?. TemplateLiteral
+                //        OptionalChain TemplateLiteral
+                // This is a hard error.
+                syntax_error("Invalid tagged template literal after ?.", position());
+                break;
+            default:
+                if (match_identifier_name())
+                    chain.append(OptionalChain::MemberReference { parse_identifier(), OptionalChain::Mode::Optional });
```

Comment:
Here you match on identifier name which can be things like `new`, `null` and other keywords.
But then you use `parse_identifier` which does not accept most of those keywords.

Although this problem is definitely worse with the multiple parse_identifier like methods we have I'm hoping to clean that up some time.

For example this should work: 
```js
const a = {
    function: {
        new: {
            null: {
                undefined: 1,
            }
        }
    }
}

console.log(a?.function?.new?.null?.undefined === 1);
console.log(a?.function?.new?.undefined?.null === undefined);
```

Reply:
I just manually made an `Identifier`, though it's kind of a trap for `parse_identifier` not to be parsing what the spec thinks of `Identifier` - perhaps we should leave a note on that?

Reply:
The problem is that there are 3? or more different concepts of "identifier" in what is allowed. And it changes with context.
Adding a note is fine but perhaps we should move to more spec like names (sorry...). But let's not do that in this PR.